### PR TITLE
Make returned arrays of gometric transforms read only

### PR DESCRIPTION
**Pull request by hmaarrfk**
This should help users accidentally avoid writing to temporary arrays

Closes https://github.com/pygfx/pygfx/issues/651

### CODE REVIEW

File: pygfx/helpers/_axes.py [hunk 1]
```diff
@@ -79,7 +79,12 @@ def __init__(self, size=1.0, thickness=2):
             arrow_head.local.position = pos
             # offset by half of height since the cones
             # are centered around the origin
-            arrow_head.local.position += arrow_size / 2 * la.vec_normalize(pos)
+            # Do not use +=
+            # see: https://github.com/pygfx/pygfx/issues/651
+            arrow_head.local.position = (
+                arrow_head.local.position +
+                arrow_size / 2 * la.vec_normalize(pos)
```
Comment (Korijn) [inline pygfx/helpers/_axes.py:86]: Well, that's an annoying and surprising side effect of this change! :/ I guess I knew that ndarrays overload this to be in-place...

File: pygfx/utils/transform.py [hunk 2]
```diff
@@ -157,12 +157,20 @@ def scaling_signs(self):
...
             # and so there is no scaling component to preserve
             # any decomposed scaling is acceptable
-            return la.mat_decompose(self.matrix)
+            decomposed = la.mat_decompose(self.matrix)
+
+        # Stop the user from accidentally writing the temporary arrays
+        # that we return from these operations
+        # https://github.com/pygfx/pygfx/issues/651
+        for m in decomposed:
+            m.flags['WRITEABLE'] = False
```
Comment (Korijn) [inline pygfx/utils/transform.py:171]: Should this flag also be set for other ndarrays in this class?

Comment (hmaarrfk) [reply pygfx/utils/transform.py:171]: I think so. I just wanted to get feedback on the potential challenges before going too far in this.

It seems that searching for usages of `la.` will find most of them.

The 730 lines of code shouldn't be too hard to go through one by one either.

**Comment by hmaarrfk** [comment]
Note that one of the arrays was already setting this flag with `flags.writeable` so I followed suit.

File: pygfx/utils/transform.py [hunk 3]
```diff
@@ -544,6 +569,9 @@ def flag_update(self):
     @property
     def matrix(self):
         view_array = self.untracked_matrix.view()
+        # Stop the user from accidentally writing the temporary arrays
+        # that we return from these operations
+        # https://github.com/pygfx/pygfx/issues/651
         view_array.flags.writeable = False
```
Comment (Korijn) [inline pygfx/utils/transform.py:575]: Does this flag only the view or does it also flag the original array?

Comment (hmaarrfk) [reply pygfx/utils/transform.py:575]: only the view:

```python
In [1]: import numpy as np

In [2]: a = np.zeros((2, 3))

In [3]: b = a.view()

In [4]: b.flags.writeable = False

In [5]: assert a.flags.writeable

In [6]: a.flags.writeable
True

In [7]: assert not b.flags.writeable

In [8]: b.flags.writeable
False
```

**Comment by Korijn** [comment]
I think this looks good to merge! @almarklein any thoughts?

**Review by Korijn — APPROVED** [review]

**Review by almarklein — APPROVED** [review]

**Pull request merged**